### PR TITLE
ARROW-14709: [C++][Java] Upgrade ORC to 1.7.1 and use the official Apache distribution site

### PR DESCRIPTION
**Pull request by williamhyun**
This PR aims to bump ORC to 1.7.1 and use the official Apache distribution site. 

Apache ORC 1.7.1 is a maintenance release with the following bug fixes.

- https://github.com/apache/orc/releases/tag/rel%2Frelease-1.7.1

### CODE REVIEW

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. The version and shasum and new URL is correct.

**Comment by dongjoon-hyun** [comment]
BTW, the CI failure error message looks weird to me. Why it try to download `ursa-labs`?
```
Connection #0 to host github.com left intact
         --- LOG END ---
         error: downloading 'https://github.com/ursa-labs/thirdparty/releases/download/latest/orc-1.7.1.tar.gz' failed
         status_code: 22
         status_string: "HTTP response code said error"
         log:
         --- LOG BEGIN ---
           Trying 140.82.114.4:443...
```

File: cpp/thirdparty/versions.txt [hunk 1]
```diff
@@ -118,7 +118,7 @@ DEPENDENCIES=(
   "ARROW_LZ4_URL lz4-${ARROW_LZ4_BUILD_VERSION}.tar.gz https://github.com/lz4/lz4/archive/${ARROW_LZ4_BUILD_VERSION}.tar.gz"
   "ARROW_MIMALLOC_URL mimalloc-${ARROW_MIMALLOC_BUILD_VERSION}.tar.gz https://github.com/microsoft/mimalloc/archive/${ARROW_MIMALLOC_BUILD_VERSION}.tar.gz"
   "ARROW_NLOHMANN_JSON_URL nlohmann-json-${ARROW_NLOHMANN_JSON_BUILD_VERSION}.tar.gz https://github.com/nlohmann/json/archive/refs/tags/${ARROW_NLOHMANN_JSON_BUILD_VERSION}.tar.gz"
-  "ARROW_ORC_URL orc-${ARROW_ORC_BUILD_VERSION}.tar.gz https://github.com/apache/orc/archive/rel/release-${ARROW_ORC_BUILD_VERSION}.tar.gz"
+  "ARROW_ORC_URL orc-${ARROW_ORC_BUILD_VERSION}.tar.gz https://archive.apache.org/dist/orc/orc-${ARROW_ORC_BUILD_VERSION}/orc-${ARROW_ORC_BUILD_VERSION}.tar.gz"
```
Comment (dongjoon-hyun) [inline cpp/thirdparty/versions.txt:121]: Could you update the following file too, @williamhyun ?
```
$ git grep ursa-labs | grep orc
cpp/cmake_modules/ThirdpartyToolchain.cmake:           "https://github.com/ursa-labs/thirdparty/releases/download/latest/orc-${ARROW_ORC_BUILD_VERSION}.tar.gz"
```

**Comment by pitrou** [comment]
@nealrichardson What is our policy wrt third-party download URLs?

**Comment by pitrou** [comment]
@nealrichardson Could you answer the question above or point to someone who would be able to give an answer?

**Comment by nealrichardson** [comment]
@pitrou Sorry, I don't understand the question.

**Comment by pitrou** [comment]
@nealrichardson How do we choose third-party download URLs/mirrors and how do we maintain them?

**Comment by nealrichardson** [comment]
`cpp/cmake_modules/ThirdpartyToolchain.cmake` needs to get updated (specifically [here](https://github.com/apache/arrow/blob/master/cpp/cmake_modules/ThirdpartyToolchain.cmake#L544)), that's where the urls are that get used by cmake (there's an old JIRA about trying to unify that and the other download script). For many dependencies we have a backup copy in https://github.com/ursa-labs/thirdparty/releases/, which can help prevent spurious build failures, though it's not required.

**Comment by pitrou** [comment]
@github-actions crossbow submit -g nightly

**Review by pitrou — APPROVED** [review]
CI seems to be green except for unrelated failures. Thank you @williamhyun !

**Comment by dongjoon-hyun** [comment]
Thank you, @pitrou , @williamhyun and all!

**Pull request closed without merging**